src/gba/GBAinline.h

```diff
@@ -165,7 +166,7 @@ static inline uint32_t CPUReadMemory(uint32_t address)
                 armMode ? armNextPC - 4 : armNextPC - 2);
         }
 #endif
-        if (cpuDmaHack) {
+        if (cpuDmaRunning || ((reg[15].I - cpuDmaPC) == (armState ? 4 : 2))) {
```

Comment:
I am a bit puzzled by what this does. I just treated myself to a crash course on GBA architecture so my understanding may be wrong.
Basically, this checks if the PC register (r15) is 0x0004 in ARM mode or 0x0002 in Thumb mode, since `cpuDmaPC` is always 0. This memory range is in the BIOS, which can't be right.

If we are missing some `cpuDmaPC` initialization - presumably to the PC location when we started the DMA operation, then this check becomes "Is PC 1 instruction ahead of the DMA start?", which makes more sense with the intent of the PR.

However, in this case, if I am reading this page properly: http://problemkaputt.de/gbatek-gba-unpredictable-things.htm, it seems that this should only be done in Thumb mode, not ARM mode, so it should be in the not `armState` branch below (and ditto below).

But we'd need to test that on real hardware to confirm

Reply:
@Squall-Leonhart does this make sense to you?

Reply:
the GBA documentation is not complete, some games take advantage of these undocumented features.

For instance, nogba has the slowdown issue, but not the hang issue that the original version of the downstream libretro patch introduced.